### PR TITLE
Spec cleanup

### PR DESCRIPTION
**Pull request by mezis**
Some spring cleaning so we can confidently refactor implementation without affecting behaviour.

- Fixes integration test flakyness (removes a [`sleep 0.5`](https://github.com/deliveroo/routemaster-drain/pull/29/files#diff-e9a141dcca8549efe3b7c9ebdcb807adL35) call)
- Fixes API client integration tests (was [testing implementation](https://github.com/deliveroo/routemaster-drain/pull/29/files#diff-812d41e9c12bedd302061adb88389c45L168))
- Fixes Cache client integration tests (was [testing implementation](https://github.com/deliveroo/routemaster-drain/pull/29/files#diff-e9a141dcca8549efe3b7c9ebdcb807adL35))
- Fixes telemetry tests (was using a [test double](https://github.com/deliveroo/routemaster-drain/pull/29/files#diff-812d41e9c12bedd302061adb88389c45L251) in an integration test)
- Adds Ruby 2.4 to the [build matrix](https://github.com/deliveroo/routemaster-drain/pull/29/files#diff-354f30a63fb0907d4ad57269548329e3R8)
- Bumps Rails versions [tested by `appraisal`](https://github.com/deliveroo/routemaster-drain/pull/29/files#diff-3bb7346aca79d8142e16e61372bd04dc) (3.2.14 → 3.2.22, 4.0.0 → 4.2.8, 5.0.0 → 5.0.1)

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -17,9 +17,9 @@ gem 'simplecov'
...
 
-group :development, :test do
-  gem 'guard-rspec',                require: false
-  gem 'pry-byebug',                 require: false
-  gem 'rspec'
-  gem 'appraisal'
-end
+gem 'guard-rspec',                require: false
+gem 'pry',                        require: false
+# gem 'pry-byebug',                 require: false
```
Comment (sideshowcoder) [inline Gemfile]: Why comment out?

File: spec/routemaster/integration/api_client_spec.rb [hunk 2]
```diff
@@ -165,70 +156,58 @@ class ProcHandler
...
+  describe 'caching behaviour' do
+    let(:url) { "#{host}/resources/1" }
+    def timestamp ; subject.get(url).body.updated_at ; end
 
-      it 'returns the response from within the future' do
-        expect(cache.hget("cache:#{url}", "v:,l:,body")).to be_nil
+    describe 'GET requests' do
+      context 'when the resource was fetched' do
+        let!(:cached_stamp) { timestamp }
+        let(:fetched_stamp) { timestamp }
```
Comment (sideshowcoder) [inline spec/routemaster/integration/api_client_spec.rb:167]: Maybe I'm confused but why the `fetched_timestamp` shouldn't just `timestamp` do for tests below, correct me if I'm wrong, but as it reads I'm a little confused why this is needed, forgive my Rspec 😸

File: spec/routemaster/integration/api_client_spec.rb [hunk 3]
```diff
@@ -242,49 +221,57 @@ class ProcHandler
...
     subject do
       Routemaster::APIClient.new(metrics_client: metrics_client,
                                  source_peer: source_peer)
     end
 
+    before do
+      wrapper = -> (m, *args, &block) {
+        calls << [m.name, *args]
+        m.call(*args, &block)
+      }
```
Comment (sideshowcoder) [inline spec/routemaster/integration/api_client_spec.rb]: As I understand it this is a spy right? https://relishapp.com/rspec/rspec-mocks/docs/basics/spies

**Review by sideshowcoder — APPROVED** [review]
Looks good, some small nitpicks but fine either way.

File: Gemfile (hunk 1 above)
Comment (mezis) [reply Gemfile]: Ah, I meant to remove it altogether. `byebug` is causing intermittent issues, and I'm not sure why we introduced it in the first place.

File: spec/routemaster/integration/api_client_spec.rb (hunk 2 above)
Comment (mezis) [reply spec/routemaster/integration/api_client_spec.rb:167]: Just naming sugar, so we can write a very expressive spec:

```ruby
it 'returns a fresh response' do
  expect(fetched_stamp).to eq(cached_stamp)
end
```

File: spec/routemaster/integration/api_client_spec.rb (hunk 3 above)
Comment (mezis) [reply spec/routemaster/integration/api_client_spec.rb]: It is. I couldn't get `rspec-mocks` to spy on a "partial double" (= real implementation with call spying), _and_ call the implementation.
I'll give it another stab though, cause it's pretty unseemly to have to reimplement this!

File: spec/routemaster/integration/api_client_spec.rb (hunk 2 above)
Comment (sideshowcoder) [reply spec/routemaster/integration/api_client_spec.rb:167]: I see, makes sense. Personally I prefer low-carb 😄  but whatever is more natural is fine with me.

**Comment by nronas** [comment]
SUPER. thanks for doing this :+1:

**Review by nronas — APPROVED** [review]

File: spec/routemaster/integration/api_client_spec.rb (hunk 3 above)
Comment (mezis) [reply spec/routemaster/integration/api_client_spec.rb]: Turns out I was too lazy, and it does work.

**Review by danielcooper — APPROVED** [review]
👍

**Pull request merged**